chain/client/src/stateless_validation/state_witness_tracker.rs

```diff
@@ -0,0 +1,201 @@
+/// message from validators to block producers to make an estimate of network time for sending
+/// witness and receiving the endorsement.
+pub struct ChunkStateWitnessTracker {
+    witnesses: LruCache<ChunkStateWitnessKey, ChunkStateWitnessRecord>,
+    clock: Clock,
+}
+
+impl ChunkStateWitnessTracker {
+    pub fn new(clock: Clock) -> Self {
+        Self { witnesses: LruCache::new(CHUNK_STATE_WITNESS_MAX_RECORD_COUNT), clock }
+    }
+
+    /// Adds a new witness message to track.
+    pub fn record_witness_sent(
+        &mut self,
+        witness: &ChunkStateWitness,
+        witness_size_in_bytes: usize,
+        num_validators: usize,
+    ) -> () {
+        let key = ChunkStateWitnessKey::new(witness);
+        tracing::debug!(target: "state_witness_tracker", "Sent state witness: {:?} of {} bytes",
+            witness.inner.chunk_header.chunk_hash(), witness_size_in_bytes);
+        self.witnesses.put(
+            key,
+            ChunkStateWitnessRecord {
+                num_validators,
+                witness_size: witness_size_in_bytes,
+                sent_timestamp: self.clock.now(),
+            },
+        );
+    }
+
+    /// Handles an ack message for the witness. Calculates the round-trip duration and
+    /// records it in the corresponding metric.
+    pub fn on_witness_ack_received(&mut self, ack: ChunkStateWitnessAck) -> () {
+        let key = ChunkStateWitnessKey { chunk_hash: ack.chunk_hash };
+        tracing::debug!(target: "state_witness_tracker", "Received ack for state witness: {:?}",
+            key.chunk_hash);
+        if let Some(mut record) = self.witnesses.pop(&key) {
+            assert!(record.num_validators > 0);
+
+            Self::update_roundtrip_time_metric(&record, &self.clock);
+
+            // Cleanup the record if we received the acks from all the validators.
+            if let Some(remaining) = record.num_validators.checked_sub(1) {
+                if remaining > 0 {
+                    record.num_validators = remaining;
+                    self.witnesses.put(key.clone(), record);
+                }
+            }
+        }
+    }
+
+    /// Records the roundtrip time in metrics.
+    fn update_roundtrip_time_metric(record: &ChunkStateWitnessRecord, clock: &Clock) -> () {
+        let received_time = clock.now();
+        if received_time > record.sent_timestamp {
+            metrics::CHUNK_STATE_WITNESS_NETWORK_ROUNDTRIP_TIME
+                .with_label_values(&[witness_size_bucket(record.witness_size)])
+                .observe((received_time - record.sent_timestamp).as_seconds_f64());
+        }
+    }
+
+    #[cfg(test)]
+    fn get_record_for_witness(
+        &mut self,
+        witness: &ChunkStateWitness,
+    ) -> Option<&ChunkStateWitnessRecord> {
+        let key = ChunkStateWitnessKey::new(witness);
+        return self.witnesses.get(&key);
+    }
+}
+
+/// Buckets for state-witness size.
+// TODO: Use size::Size to represent sizes.
```

Comment:
not sure what exaclty is happening but bytesize is another utility that is commonly used in the codebase